docs/topics/object-declarations.md

```diff
@@ -193,32 +193,92 @@ fun main() {
 ```kotlin
-data object MyObject
+data object MyDataObject {
+    val x: Int = 3
+}
 
 fun main() {
-    println(MyObject) // MyObject
+    println(MyDataObject) // MyDataObject
 }
 ```
 
-[Sealed class hierarchies](sealed-classes.md) are a particularly good fit for `data object` declarations, since they allow you to maintain symmetry with any data classes you might have defined alongside the object:
+The `equals()` function for a `data object` ensures that all objects that have the type of your `data object` are considered equal.
+In most cases, you will only have a single instance of your data object at runtime (after all, a `data object` declares a singleton).
+However, in the edge case where another object of the same type is generated at runtime (e.g. via platform reflection through `java.lang.reflect`, or by using a JVM serialization library that uses this API under the hood), this ensures that the objects are treated as being equal.
+
+> Make sure to only compare `data objects` structurally (i.e. using the `==` operator), never by reference (the `===` operator). This helps you avoid pitfalls when more than once instance of a data object exists at runtime.
+{type="warning"}
 
 ```kotlin
-sealed class ReadResult {
-    data class Number(val value: Int): ReadResult()
-    data class Text(val value: String): ReadResult()
-    data object EndOfFile: ReadResult()
+import java.lang.reflect.Constructor
+
+data object MySingleton
+
+fun main() {
+    val evilTwin = createInstanceViaReflection()
+
+    println(MySingleton) // MySingleton
+    println(evilTwin) // MySingleton
+
+    // Even when a library forcefully creates a second instance of MySingleton, its `equals` method returns true:
+    println(MySingleton == evilTwin) // true
+
+    // Do not compare data objects via ===.
+    println(MySingleton === evilTwin) // false
+}
+
+fun createInstanceViaReflection(): MySingleton {
+    // Kotlin reflection does not permit the instantiation of data objects.
+    // This creates a new MySingleton instance "by force" (i.e. Java platform reflection)
+    // Don't do this yourself!
+    return (MySingleton.javaClass.declaredConstructors[0].apply { isAccessible = true } as Constructor<MySingleton>).newInstance()
+}
+```
+
+The generated `hashCode()` function has behavior that is consistent with the `equals()` function, so that all runtime instances of a `data object` have the same hash code.
+
+#### No copy and componentN functions
+
+While `data object` and `data class` declarations are often used together and have some similarities, there are some functions that are not generated for a `data object`:
+
+- No `copy()` function: Because a `data object` declaration is intended to be used as singleton objects, no `copy()` function is generated. The singleton pattern restricts the instantiation of a class to one, single instance, which would be violated by allowing copies of the instance to be created.
```

Comment:
```suggestion
* No `copy()` function. Because a `data object` declaration is intended to be used as singleton objects, no `copy()` function is generated. The singleton pattern restricts the instantiation of a class to a single instance, which would be violated by allowing copies of the instance to be created.
```